### PR TITLE
feat: add date-fns and Moment.js to Temporal API migration recipes

### PR DESCRIPTION
**Pull request by zieka**
## Summary
- Adds two new JavaScript migration recipes for moving from date libraries to the native Temporal API
- **`DateFnsToTemporal`** — migrates 18 core date-fns functions: arithmetic (`addDays/Months/Years`, `subDays/Months/Years`), comparison (`isAfter`, `isBefore`, `isEqual`), difference (`differenceInDays/Hours/Minutes`), and period boundaries (`startOfDay/Month/Year`, `endOfDay/Month/Year`)
- **`MomentToTemporal`** — migrates Moment.js patterns: creation (`moment()`, `moment(x)`, `moment.utc()`), arithmetic (`.add()`/`.subtract()` with full/shorthand/singular units including milliseconds), comparison (`.isBefore()`/`.isAfter()`/`.isSame()`), `.clone()`, `.startOf()`, and method chains
- Uses declarative tagged template API (`pattern`/`template`/`raw()`) with capture constraints and `.orElse()` rule composition
- Supports aliased named imports (e.g., `import { addDays as plusDays }`)
- All `startOf`/`endOf` functions return consistent `PlainDateTime` types matching the original library semantics
- Both recipes registered in the marketplace under `JavaScript > Migrate > date-fns` and `JavaScript > Migrate > Moment.js`

## Test plan
- [x] 21 date-fns tests covering all 18 functions, guard, multi-import, and aliased import edge cases
- [x] 21 Moment.js tests covering creation, arithmetic (including singular units + milliseconds), comparison, clone, startOf, and chains
- [x] Full migrate test suite passes (98/98 across 6 suites)
- [x] `startOf`/`endOf` functions return PlainDateTime consistently
- [x] Marketplace registration for both recipes
- [x] Refactored to declarative tagged template API per reviewer feedback
- [ ] Review generated Temporal output for semantic correctness in real-world usage
- [ ] Consider follow-up for `format`/`parse`, deep import paths (`date-fns/addDays`), and Moment locale support

### CODE REVIEW

File: rewrite-javascript/rewrite/test/javascript/migrate/date-fns/date-fns-to-temporal.test.ts [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    // --- Multiple imports ---
+
+    test("multiple date-fns imports", () => {
+        return spec.rewriteRun(
+            //language=typescript
+            typescript(
+                `import { addDays, subMonths } from 'date-fns';\nconst a = addDays(d, 1);\nconst b = subMonths(d, 2);`,
+                `const a = Temporal.PlainDate.from(d).add({days: 1});\nconst b = Temporal.PlainDate.from(d).subtract({months: 2});`
+            )
+        );
```
[the remark this answers is not in the document]
Comment (zieka) [reply rewrite-javascript/rewrite/test/javascript/migrate/date-fns/date-fns-to-temporal.test.ts:236]: Done — added a test for aliased imports (`import { addDays as plusDays }`) in 84a15d59b.

File: rewrite-javascript/rewrite/src/javascript/migrate/date-fns/date-fns-to-temporal.ts [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+        if (!importClause?.namedBindings) return;
+        if (importClause.namedBindings.kind !== JS.Kind.NamedImports) return;
+
+        const namedImports = importClause.namedBindings as JS.NamedImports;
+        for (const elem of namedImports.elements.elements) {
+            const specifier = elem.element;
+            const specifierNode = specifier.specifier;
+            if (isIdentifier(specifierNode) && DATE_FNS_FUNCTIONS.has(specifierNode.simpleName)) {
+                this.importedDateFnsFunctions.add(specifierNode.simpleName);
+            }
```
[the remark this answers is not in the document]
Comment (zieka) [reply rewrite-javascript/rewrite/src/javascript/migrate/date-fns/date-fns-to-temporal.ts:105]: Done — added `JS.Kind.Alias` handling in `collectDateFnsImports()` that tracks a `aliasToOriginal` map. The local name is used for call-site matching and the original name for import removal. See 84a15d59b.

File: rewrite-javascript/rewrite/src/javascript/migrate/date-fns/date-fns-to-temporal.ts [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+            .capture(b)
+            .code(')')
+            .build();
+
+        const tmpl = Template.builder()
+            .code('Temporal.PlainDateTime.compare(')
+            .param(a)
+            .code(', ')
+            .param(b)
+            .code(`) ${operator}`)
```
[the remark this answers is not in the document]
Comment (zieka) [reply rewrite-javascript/rewrite/src/javascript/migrate/date-fns/date-fns-to-temporal.ts]: Intentionally not addressing this. This recipe assumes a migration context where the developer is moving their codebase to Temporal — generating runtime type-checking code like `new Date(...).toISOString()` wrappers would produce noisy output that the developer would need to clean up anyway. The simpler output is easier to review and adjust during migration.

File: rewrite-javascript/rewrite/src/javascript/migrate/date-fns/date-fns-to-temporal.ts [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+            .capture(b)
+            .code(')')
+            .build();
+
+        // date-fns differenceIn*(a, b) computes a - b, so use a.since(b)
+        const tmpl = Template.builder()
+            .param(a)
+            .code(`.since(`)
+            .param(b)
+            .code(`, {largestUnit: "${largestUnit}"}).${field}`)
```
[the remark this answers is not in the document]
Comment (zieka) [reply rewrite-javascript/rewrite/src/javascript/migrate/date-fns/date-fns-to-temporal.ts]: Same rationale as the compare comment — the recipe produces clean Temporal API calls and assumes the migration context will handle input type alignment. Generating `typeof x === "number" ? x : x.getTime()` wrappers would be overly defensive for a rewrite recipe.

File: rewrite-javascript/rewrite/src/javascript/migrate/date-fns/date-fns-to-temporal.ts [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+            case 'differenceInDays': return this.differenceRule('differenceInDays', 'day', 'days');
+            case 'differenceInHours': return this.differenceRule('differenceInHours', 'hour', 'hours');
+            case 'differenceInMinutes': return this.differenceRule('differenceInMinutes', 'minute', 'minutes');
+
+            case 'startOfDay': return this.singleArgRule('startOfDay', 'Temporal.PlainDate.from(', ').toPlainDateTime()');
+            case 'startOfMonth': return this.singleArgRule('startOfMonth', 'Temporal.PlainDate.from(', ').with({ day: 1 })');
+            case 'startOfYear': return this.singleArgRule('startOfYear', 'Temporal.PlainDate.from(', ').with({ month: 1, day: 1 })');
+            case 'endOfDay': return this.singleArgRule('endOfDay', 'Temporal.PlainDate.from(', ').toPlainDateTime({hour: 23, minute: 59, second: 59, millisecond: 999})');
+            case 'endOfMonth': return this.endOfMonthRule();
+            case 'endOfYear': return this.singleArgRule('endOfYear', 'Temporal.PlainDate.from(', ').with({ month: 12, day: 31 })');
```
[the remark this answers is not in the document]
Comment (zieka) [reply rewrite-javascript/rewrite/src/javascript/migrate/date-fns/date-fns-to-temporal.ts]: Done — `startOfMonth`/`startOfYear` now append `.toPlainDateTime()` to zero time components, and `endOfYear` now includes `.toPlainDateTime({hour: 23, minute: 59, second: 59, millisecond: 999})`. See 84a15d59b.

File: rewrite-javascript/rewrite/src/javascript/migrate/date-fns/index.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+import {CategoryDescriptor} from "../../../marketplace";
+import {Migrate} from "../index";
+
+export {DateFnsToTemporal} from "./date-fns-to-temporal";
+
+export const MigrateDateFns: CategoryDescriptor[] = [...Migrate, {
+    displayName: "date-fns",
+    description: "Migrate from date-fns to the native Temporal API."
+}]
```
[the remark this answers is not in the document]
Comment (zieka) [reply rewrite-javascript/rewrite/src/javascript/migrate/date-fns/index.ts:25]: Done — both `DateFnsToTemporal` and `MomentToTemporal` are now registered in `activate()` in `src/index.ts`. See 84a15d59b.

File: rewrite-javascript/rewrite/src/javascript/migrate/moment/moment-to-temporal.ts [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+const UNIT_MAP: Record<string, string> = {
+    'years': 'years', 'y': 'years',
+    'months': 'months', 'M': 'months',
+    'weeks': 'weeks', 'w': 'weeks',
+    'days': 'days', 'd': 'days',
+    'hours': 'hours', 'h': 'hours',
+    'minutes': 'minutes', 'm': 'minutes',
+    'seconds': 'seconds', 's': 'seconds',
+    'milliseconds': 'milliseconds', 'ms': 'milliseconds',
+};
```
[the remark this answers is not in the document]
Comment (zieka) [reply rewrite-javascript/rewrite/src/javascript/migrate/moment/moment-to-temporal.ts:34]: Done — added all singular forms (`'day'`, `'month'`, `'year'`, `'week'`, `'hour'`, `'minute'`, `'second'`, `'millisecond'`) to `UNIT_MAP` with a test for `.add(1, 'day')`. See fadfc3502.

File: rewrite-javascript/rewrite/src/javascript/migrate/date-fns/date-fns-to-temporal.ts [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+
+        const pat = Pattern.builder()
+            .code(`${fnName}(`)
+            .capture(date)
+            .code(')')
+            .build();
+
+        // Use IIFE to avoid referencing .daysInMonth on the unconverted input
+        const tmpl = Template.builder()
+            .code('((d) => d.with({day: d.daysInMonth}))(Temporal.PlainDate.from(')
```
[the remark this answers is not in the document]
Comment (zieka) [reply rewrite-javascript/rewrite/src/javascript/migrate/date-fns/date-fns-to-temporal.ts]: Done — `endOfMonth` IIFE now appends `.toPlainDateTime({hour: 23, minute: 59, second: 59, millisecond: 999})`, consistent with `endOfDay` and `endOfYear`. See fadfc3502.

File: rewrite-javascript/rewrite/test/javascript/migrate/date-fns/date-fns-to-temporal.test.ts [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+    }, 60000);
+
+    test("endOfMonth", () => {
+        return spec.rewriteRun(
+            //language=typescript
+            typescript(
+                `import { endOfMonth } from 'date-fns';\nconst result = endOfMonth(date);`,
+                `const result = ((d) => d.with({day: d.daysInMonth}))(Temporal.PlainDate.from(date));`
+            )
+        );
```
[the remark this answers is not in the document]
Comment (zieka) [reply rewrite-javascript/rewrite/test/javascript/migrate/date-fns/date-fns-to-temporal.test.ts:203]: Done — test updated to expect PlainDateTime output. See fadfc3502.

File: rewrite-javascript/rewrite/src/javascript/migrate/moment/moment-to-temporal.ts [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+                if (effectiveArgs.length === 1) {
+                    // Note: Temporal.Instant.from() requires an ISO 8601 string with offset/timezone;
+                    // moment.utc(x) accepts more formats, so this may need manual adjustment
+                    const arg = capture('arg');
+                    return await Template.builder()
+                        .code('Temporal.Instant.from(')
+                        .param(arg)
+                        .code(')')
+                        .build()
+                        .apply(method, this.cursor, {values: new Map([[arg, effectiveArgs[0].element]])});
```
[the remark this answers is not in the document]
Comment (zieka) [reply rewrite-javascript/rewrite/src/javascript/migrate/moment/moment-to-temporal.ts]: The recipe already has an inline comment documenting this constraint. Restricting rewrites to only string literals with offsets would be overly narrow — in practice, `moment.utc(x)` is often called with variables whose values are offset-bearing at runtime. The recipe intentionally produces the closest Temporal equivalent, and developers can adjust during review. This is a known limitation documented in the PR description's follow-up section.

File: rewrite-javascript/rewrite/test/javascript/migrate/moment/moment-to-temporal.test.ts [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+            )
+        );
+    }, 60000);
+
+    test("moment.utc(string)", () => {
+        return spec.rewriteRun(
+            //language=typescript
+            typescript(
+                `import moment from 'moment';\nconst d = moment.utc('2024-01-15');`,
+                `const d = Temporal.Instant.from('2024-01-15');`
```
[the remark this answers is not in the document]
Comment (zieka) [reply rewrite-javascript/rewrite/test/javascript/migrate/moment/moment-to-temporal.test.ts:73]: See reply on the source comment — we're keeping the current behavior with the inline comment as documentation.

File: rewrite-javascript/rewrite/src/javascript/migrate/moment/moment-to-temporal.ts [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+                    .apply(method, this.cursor, {
+                        values: new Map([[obj, method.select.element]])
+                    });
+            }
+        }
+
+        // .startOf(unit)
+        if (methodName === 'startOf' && method.select) {
+            return this.transformStartOf(method);
+        }
```
[the remark this answers is not in the document]
Comment (zieka) [reply rewrite-javascript/rewrite/src/javascript/migrate/moment/moment-to-temporal.ts]: Valid edge case. In practice, `moment.utc().startOf('day')` is uncommon — UTC instant creation is typically used for timestamps, not date manipulation. The recipe handles the common patterns; mixed chains involving `moment.utc()` with date methods would need manual adjustment. This could be addressed in a follow-up with type-aware chain analysis.

File: rewrite-javascript/rewrite/src/javascript/migrate/moment/moment-to-temporal.ts [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+            .param(b)
+            .code(`) ${operator}`)
+            .build()
+            .apply(method, this.cursor, {
+                values: new Map([
+                    [a, method.select!.element],
+                    [b, args[0].element]
+                ])
+            });
+    }
```
[the remark this answers is not in the document]
Comment (zieka) [reply rewrite-javascript/rewrite/src/javascript/migrate/moment/moment-to-temporal.ts]: Same migration-context reasoning as the earlier round — the recipe transforms the call shape and assumes the developer is migrating their entire codebase to Temporal types. Generating defensive wrappers would produce noisy output.

File: rewrite-javascript/rewrite/src/javascript/migrate/date-fns/date-fns-to-temporal.ts [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+            .code(')')
+            .build();
+
+        const tmpl = Template.builder()
+            .code(`Temporal.PlainDate.from(`)
+            .param(date)
+            .code(`).${method}({ ${unit}: `)
+            .param(amount)
+            .code(' })')
+            .build();
```
[the remark this answers is not in the document]
Comment (zieka) [reply rewrite-javascript/rewrite/src/javascript/migrate/date-fns/date-fns-to-temporal.ts]: This is an intentional design choice. `date-fns` arithmetic functions (`addDays`, `subMonths`, etc.) conceptually operate on calendar dates. `Temporal.PlainDate` is the idiomatic Temporal type for this. While date-fns accepts/returns `Date` objects (which carry time), the Temporal API correctly separates date-only from date-time concerns. If the user needs time preservation, `PlainDateTime` variants could be added as a follow-up recipe option.

File: rewrite-javascript/rewrite/test/javascript/migrate/date-fns/date-fns-to-temporal.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (knutwannheden) [reply rewrite-javascript/rewrite/test/javascript/migrate/date-fns/date-fns-to-temporal.test.ts:236]: Hmm. I actually thought that if your patterns are properly type attributed, the matching would "see through" these aliases without you having to do any extra work. Worth a try maybe.

**Comment by zieka** [comment]
continuing work in different repo
